realm/src/main/java/io/realm/DynamicRealm.java

```diff
@@ -0,0 +1,283 @@
+                                                                    Sort sortOrder1, String fieldName2,
+                                                                    Sort sortOrder2) {
+        return allObjectsSorted(className, new String[]{fieldName1, fieldName2}, new Sort[]{sortOrder1,
+                sortOrder2});
+    }
+
+    /**
+     * Get all objects of a specific class name sorted by multiple fields.  If no objects exist, the
+     * returned RealmResults will not be null. Use {@link RealmResults#size()} to check the number of
+     * objects instead.
+     *
+     * @param className The class to get all objects from.
+     * @param sortOrders sort ascending if SORT_ORDER_ASCENDING, sort descending if SORT_ORDER_DESCENDING.
+     * @param fieldNames an array of field names to sort objects by.
+     *        The objects are first sorted by fieldNames[0], then by fieldNames[1] and so forth.
+     * @return A sorted RealmResults containing the objects.
+     * @throws java.lang.IllegalArgumentException if a field name does not exist.
+     */
+    @SuppressWarnings("unchecked")
+    public RealmResults<DynamicRealmObject> allObjectsSorted(String className, String fieldNames[],
+                                                                    Sort sortOrders[]) {
+        checkAllObjectsSortedParameters(fieldNames, sortOrders);
+        Table table = this.getTable(className);
+        TableView tableView = doMultiFieldSort(fieldNames, sortOrders, table);
+
+        return RealmResults.createFromDynamicQuery(this, tableView, className);
+    }
+
+    private static synchronized DynamicRealm create(RealmConfiguration configuration) {
+
+        // Check if a cached instance already exists for this thread
+        String canonicalPath = configuration.getPath();
+        Map<RealmConfiguration, Integer> localRefCount = referenceCount.get();
+        Integer references = localRefCount.get(configuration);
+        if (references == null) {
+            references = 0;
+        }
+        Map<RealmConfiguration, DynamicRealm> realms = realmsCache.get();
+        DynamicRealm realm = realms.get(configuration);
+        if (realm != null) {
+            localRefCount.put(configuration, references + 1);
+            return realm;
+        }
+
+        // Create new DynamicRealm and cache it. All exception code paths must close the DynamicRealm otherwise we risk
+        // serving faulty cache data.
+        validateAgainstExistingConfigurations(configuration);
+        boolean autoRefresh = Looper.myLooper() != null;
+        realm = new DynamicRealm(configuration, autoRefresh);
+        List<RealmConfiguration> pathConfigurationCache = globalPathConfigurationCache.get(canonicalPath);
+        if (pathConfigurationCache == null) {
+            pathConfigurationCache = new CopyOnWriteArrayList<RealmConfiguration>();
+            globalPathConfigurationCache.put(canonicalPath, pathConfigurationCache);
```

Comment:
This would be a problem. `globalPathConfigurationCache` is a static member on `BaseRealm` and lock of this function is on `DynamicRealm`. 

It would be logically simpler that we just have lock on `BaseRealm.class` for all `BaseRealm`, `Realm` and `DynamicRealm`.